command/upload.go

```diff
@@ -99,12 +115,172 @@ func (cmd *Upload) DoRun(args []string) (err error) {
 		HandleError(err, cmd.opts.VerboseOutput)
 	}
 
-	err = client.UploadDir(dataPath, ds.Root, &stdoutkw{}, 64)
+	size, err := totalTarSize(dataPath)
 	if err != nil {
 		HandleError(err, cmd.opts.VerboseOutput)
 	}
 
-	logrus.Infof("Uploaded data to dataset with ID: %v", ds.DatasetID)
+	header := &data.MetadataHeader{
+		Size:    size,
+		Created: time.Now(),
+		Updated: time.Now(),
+	}
+	metadata := data.NewMetadata(header, data.NewBufferedKeyReadWiter())
+	doneCh := make(chan error)
+	progressCh := make(chan int64)
+	progressBarDoneCh := make(chan struct{})
+	pr, pw := io.Pipe()
+
+	if !cmd.opts.JSONOutput {
+		go ProgressBar(size, progressCh, progressBarDoneCh)
+	} else {
+		go func() {
+			for _ = range progressCh {
+			}
+			progressBarDoneCh <- struct{}{}
+		}()
+	}
+	go func() {
+		defer close(progressCh)
+		err := client.ChunkedUpload(pr, metadata, UploadConcurrency, ds.Root, progressCh)
+		pr.Close()
+		doneCh <- err
+	}()
+
+	err = tardir(dataPath, pw)
+	if err != nil && errors.Cause(err) != io.ErrClosedPipe {
+		HandleError(errors.Wrapf(err, "failed to tar '%s'", dataPath), cmd.opts.VerboseOutput)
+	}
+
+	pw.Close()
+	err = <-doneCh
+	<-progressBarDoneCh
+	if err != nil {
+		HandleError(errors.Wrapf(err, "failed to upload '%s'", dataPath), cmd.opts.VerboseOutput)
+	}
+
+	metastring, err := metadata.ToString()
+	if err != nil {
+		HandleError(errors.Wrap(err, "failed to convert metadata to string"), cmd.opts.VerboseOutput)
+	}
+	err = client.Upload(path.Join(ds.Root, data.MetadataObjectKey), strings.NewReader(metastring))
+	if err != nil {
+		return errors.Wrap(err, "failed to upload index file")
+	}
 
 	return nil
 }
+
+//tar archives the given directory and writes bytes to w.
+func tardir(dir string, w io.Writer) (err error) {
+	tw := tar.NewWriter(w)
+	err = filepath.Walk(dir, func(path string, fi os.FileInfo, err error) error {
+		if fi.Mode().IsDir() {
+			return nil
+		}
+
+		rel, err := filepath.Rel(dir, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to determine path '%s' relative to '%s'", path, dir)
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open file '%s'", rel)
+		}
+
+		err = tw.WriteHeader(&tar.Header{
+			Name:    rel,
+			Mode:    int64(fi.Mode()),
+			ModTime: fi.ModTime(),
```

Comment:
I think we do. The header will only be a few chunks. Invalidating these chunks is worth the up-to-date information in the archive header.